NDTensors/src/lib/TypeParameterAccessors/src/typeparamaccessor.jl

```diff
@@ -0,0 +1,23 @@
+# Relies on Julia internals
+Base.@assume_effects :terminates_globally to_datatype(type::UnionAll) =
+  Base.unwrap_unionall(type)
+Base.@assume_effects :terminates_globally to_unionall(
+  type::DataType, unionall_reference::Type
+) = Base.rewrap_unionall(type, unionall_reference)
+
+# `TypeParameterAccessor` functionality.
+to_datatype(type::DataType) = type
+to_unionall(type::UnionAll, unionall_reference::Type) = type
+to_unionall(type::UnionAll) = type
+to_unionall(type::DataType) = to_unionall(type, unspecify_parameters(type))
+
+# `TypeParameterAccessor` functionality.
+to_unionall(type::Type) = to_unionall(type, unspecify_parameters(type))
```

Comment:
I wonder if we can just remove these given the new `set_parameter` implementation. We could just use `Base.unwrap_unionall` and `Base.rewrap_unionall` as needed.

Reply:
I tried re-implementing these  into the code to try and help areas I know still have type stability issues (i.e. when there is a `Core.Compiler.PartialTypeVar`) but they did not help so I am going to remove this code for now